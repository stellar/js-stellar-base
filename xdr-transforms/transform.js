import * as dom from 'dts-dom';
import fs from 'fs';
import xdrInt from './integer';
import enumToTS from './enum';

function isXDRMemberCall(node) {
  return node.type === 'MemberExpression' && node.object.name === 'xdr';
}

function isXDRIntOrUint(node) {
  return (
    node.type === 'Identifier' && (node.name === 'uint' || node.name === 'int')
  );
}

function typeDef(api, node, ns, xdrTypes) {
  const [literal, exp] = node.arguments;
  const name = literal.value;

  if (exp.type === 'CallExpression') {
    if (isXDRMemberCall(exp.callee)) {
      if (isXDRIntOrUint(exp.callee.property)) {
        switch (exp.callee.property.name) {
          case 'int':
            ns.members.push(
              dom.create.const(
                name,
                xdrTypes.INT,
                dom.DeclarationFlags.ReadOnly
              )
            );
            break;
          case 'uint':
            ns.members.push(
              dom.create.const(
                name,
                xdrTypes.UINT,
                dom.DeclarationFlags.ReadOnly
              )
            );
            break;
        }
      }
    }
  }
}

export default function transformer(file, api) {
  const j = api.jscodeshift;
  const types = j.types;
  const xdrDefs = j(file.source).findVariableDeclarators('types');

  const ns = dom.create.namespace('xdr');

  const signedInt = xdrInt(
    ns,
    'SignedInteger',
    Math.pow(2, 31) - 1,
    -Math.pow(2, 31)
  );
  const unsignedInt = xdrInt(ns, 'UnsignedInteger', Math.pow(2, 32) - 1, 0);

  ns.members.push(signedInt);
  ns.members.push(unsignedInt);

  const xdrTypes = {
    INT: signedInt,
    UINT: unsignedInt
  };

  xdrDefs.find(types.namedTypes.CallExpression).forEach((p) => {
    const node = p.value;
    const callee = node.callee;

    if (callee.type === 'MemberExpression' && callee.object.name === 'xdr') {
      const xdrType = callee.property.name;

      switch (xdrType) {
        case 'enum':
          enumToTS(api, node.arguments, ns, types);
        case 'typedef':
          typeDef(api, node, ns, xdrTypes);
        default:
          break;
      }
    }
  });

  // Use this to output to a file instead of overriding source
  if (process.env.OUT) {
    console.log('writing to: ' + process.env.OUT);
    fs.writeFileSync(process.env.OUT, dom.emit(ns));
  } else {
    return dom.emit(ns);
  }
}
