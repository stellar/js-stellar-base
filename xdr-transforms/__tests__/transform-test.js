import { runSnapshotTest } from 'jscodeshift/dist/testUtils';
import transform from '../transform';
import path from 'path';
import fs from 'fs';

jest.autoMockOff();

test.each([
  ['xdr enum', 'enum.js', {}],
  ['type definitions', 'typedef.js', {}],
  ['xdr structs', 'struct.js', {}],
  ['xdr unions', 'union.js', {}]
])('.transform(%s)', (desc, source, opts) => {
  const fixtures = path.join(__dirname, '..', '__testfixtures__');
  const content = fs.readFileSync(path.join(fixtures, source)).toString();

  runSnapshotTest(transform, opts, {
    source: content
  });
});
