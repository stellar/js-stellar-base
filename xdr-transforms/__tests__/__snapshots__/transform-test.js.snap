// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.transform(type definitions) 1`] = `
"declare namespace xdr {
    interface SignedInteger {
        readonly MAX_VALUE: 2147483647;
        readonly MIN_VALUE: -2147483648;
        read(io: Buffer): number;
        write(value: number, io: Buffer): void;
        isValid(value: number): boolean;
        toXDR(value: number): Buffer;
        fromXDR(input: Buffer, format?: \\"raw\\"): number;
        fromXDR(input: string, format: \\"hex\\" | \\"base64\\"): number;
    }

    interface UnsignedInteger {
        readonly MAX_VALUE: 4294967295;
        readonly MIN_VALUE: 0;
        read(io: Buffer): number;
        write(value: number, io: Buffer): void;
        isValid(value: number): boolean;
        toXDR(value: number): Buffer;
        fromXDR(input: Buffer, format?: \\"raw\\"): number;
        fromXDR(input: string, format: \\"hex\\" | \\"base64\\"): number;
    }

    const Uint32: UnsignedInteger;

    const Int32: SignedInteger;

}"
`;

exports[`.transform(xdr enum) 1`] = `
"declare namespace xdr {
    interface SignedInteger {
        readonly MAX_VALUE: 2147483647;
        readonly MIN_VALUE: -2147483648;
        read(io: Buffer): number;
        write(value: number, io: Buffer): void;
        isValid(value: number): boolean;
        toXDR(value: number): Buffer;
        fromXDR(input: Buffer, format?: \\"raw\\"): number;
        fromXDR(input: string, format: \\"hex\\" | \\"base64\\"): number;
    }

    interface UnsignedInteger {
        readonly MAX_VALUE: 4294967295;
        readonly MIN_VALUE: 0;
        read(io: Buffer): number;
        write(value: number, io: Buffer): void;
        isValid(value: number): boolean;
        toXDR(value: number): Buffer;
        fromXDR(input: Buffer, format?: \\"raw\\"): number;
        fromXDR(input: string, format: \\"hex\\" | \\"base64\\"): number;
    }

    class ChangeTrustResultCode {
        readonly name: \\"changeTrustMalformed\\" | \\"foo\\";

        readonly value: -1 | 1;

        static changeTrustMalformed(): ChangeTrustResultCode;

        static foo(): ChangeTrustResultCode;

    }

}"
`;
